### ISSUE TITLE
Nested calls to IMediator.Send stall out and do not return.

### ISSUE THREAD

**Issue by jkim004s** [jbogard/MediatR#285]
In the previous version of MediatR we had the ability to have nested calls to IMediator.Send method without issue. We discovered that the new version just stalls out and never returns.

Example:
Call to IMediator.Send with "MyRequest"
Handled by "MyRequestValidator" (pipeline validation behavior) 
"MyRequestValidator" makes call to "MySecurityService"
"MySecurityService" makes a call to IMediator.Send with "MySecurityRequest"
"MySecurityRequest" gets handled successfully but never returns to "MySecurityService"s call to mediator and stalls out.

This was working fine before the upgrade to the new version, but is somehow no longer working.
We are not able to revert to the previous version as the upgrade to the new version required lots of refactoring and on top of that we have released several service packs to our app after the upgrade.

**Comment by jbogard** [comment]
Do you have a more complete example?

**Comment by jkim004s** [comment]
It will take some time but I think I can clip together some of the highlights for you. Will get back as soon as I can...

**Comment by jkim004s** [comment]
Jimmy, I set up a bunch of unit tests to verify the issue and it looks like @remcoros is correct, we do have a deadlocking issue in our "UnitOfWork" pipeline behavior. Looks like we had concurrent connection deadlocking.

**Comment by VictorYMH** [comment]
Hi Jimmy, does the nested call in MediatR have any negative influence on program, should we avoid that? thanks in advace :)    @jbogard

**Comment by jbogard** [comment]
@VictorYMH yes - we found that it got too confusing, especially with pipelines. Instead, just use standard refactoring techniques for any kind of shared logic.

**Comment by VictorYMH** [comment]
So the nested call won't slow down the program dramatically, right? if pipeline is simple and effecient @jbogard

**Comment by jbogard** [comment]
@VictorYMH No it's not about slowing things down, it's about the confusion it brings. We created separate sorts of encapsulations/abstractions for inside handlers to make sure the MediatR requests/handlers were the top level of encapsulation.
